packages/core/src/authorization.ts

```diff
@@ -0,0 +1,28 @@
+import {Ctx} from "./middleware"
+import {AuthenticatedSessionContext} from "./supertokens"
+import {ZodSchema, infer as zInfer} from "zod"
+
+export type ProtectArgs<T> = {schema?: T; authorize?: boolean | unknown}
+
+interface AuthenticatedCtx extends Ctx {
+  session: AuthenticatedSessionContext
+}
+
+export const protect = <T extends ZodSchema<any, any>, U = zInfer<T>>(
+  {schema, authorize = true}: ProtectArgs<T>,
+  resolver: (args: U, ctx: AuthenticatedCtx) => any,
+) => {
+  return (rawInput: U, ctx: Ctx) => {
```

Comment:
Did you not run into typing issues client-side without typing the return value? I had issues leaving it like this and had to introduce another generic type V that extends the resolver function type and then typed the return value of the returned function as:

return (rawInput: U, ctx: Ctx): ReturnType<V> => {